### ISSUE TITLE
Propagate cluster_axis from shlo to ttnn in all_gather

### ISSUE THREAD

**Issue by tapspatel** [tenstorrent/tt-mlir#2061]
Currently, the all_gather API is hardcoded to using cluster axis 1 to perform the all_gather concatenation. Extend this to be dynamic for the mesh shape that arrives from shlo.

example test case

```
module @jit_fwd attributes {mhlo.num_partitions = 8 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main(%arg0: tensor<8192x784xf32>) -> (tensor<32768x784xf32> {jax.result_info = ""}) {
    %0 = stablehlo.custom_call @Sharding(%arg0) {backend_config = "", mhlo.sharding = "{devices=[4,1,2]<=[2,4]T(1,0) last_tile_dim_replicate}"} : (tensor<8192x784xf32>) -> tensor<8192x784xf32>
    %1 = stablehlo.custom_call @SPMDFullToShardShape(%0) {backend_config = "", mhlo.sharding = "{manual}"} : (tensor<8192x784xf32>) -> tensor<2048x784xf32>
    %2 = call @shmap_body(%1) : (tensor<2048x784xf32>) -> tensor<8192x784xf32>
    %3 = stablehlo.custom_call @Sharding(%2) {backend_config = "", mhlo.sharding = "{manual}"} : (tensor<8192x784xf32>) -> tensor<8192x784xf32>
    %4 = stablehlo.custom_call @SPMDShardToFullShape(%3) {backend_config = "", mhlo.sharding = "{devices=[4,1,2]<=[2,4]T(1,0) last_tile_dim_replicate}"} : (tensor<8192x784xf32>) -> tensor<32768x784xf32>
    return %4 : tensor<32768x784xf32>
  }
  func.func private @shmap_body(%arg0: tensor<2048x784xf32>) -> (tensor<8192x784xf32> {jax.result_info = "[('batch',), None]"}) {
    %0 = "stablehlo.all_gather"(%arg0) <{all_gather_dim = 0 : i64, channel_handle = #stablehlo.channel_handle<handle = 1, type = 1>, replica_groups = dense<[[0, 1, 2, 3], [4, 5, 6, 7]]> : tensor<2x4xi64>, use_global_device_ids}> : (tensor<2048x784xf32>) -> tensor<8192x784xf32>
    return %0 : tensor<8192x784xf32>
  }
}
```

This would result if we open our mesh as the following, which is a valid test case in llmbox (ie data parallelism across 4 devices, with model parallelism across 2 devices)

```
mesh = jax.make_mesh((2, 4), ('model', 'batch'))
```

**Comment by tapspatel** [comment]
Merged: bc853e1f14e250a179360a86858bf12d299d2777
